File: android/bmap/src/main/kotlin/dev/openbose/bmap/BmapPacket.kt
package dev.openbose.bmap

/**
 * Pure Kotlin BMAP frame codec: unrelated to Android's Bluetooth transport and
 * deliberately incapable of writing to hardware by itself.
 */
class BmapPacket(
    val block: Int,
    val function: Int,
    val operator: Int,
    payload: ByteArray = byteArrayOf()
) {
    private val bytes = payload.copyOf()
    val payload: ByteArray get() = bytes.copyOf()

    init {
        require((block in 0..255) && (function in 0..255) && (operator in 0..255))
        require(bytes.size <= 255)
    }

    fun toWire(): ByteArray =
        byteArrayOf(block.toByte(), function.toByte(), operator.toByte(), bytes.size.toByte()) + bytes

    companion object {
        const val GET = 0x01
        const val SETGET = 0x02
        const val STATUS = 0x03
        const val ERROR = 0x04
        const val START = 0x05

        fun parseExact(wire: ByteArray): BmapPacket {
            require(wire.size >= 4) { "Incomplete BMAP header" }
            val expected = 4 + (wire[3].toInt() and 0xFF)
            require(wire.size == expected) { "BMAP length mismatch: expected $expected, received ${wire.size}" }
            return BmapPacket(wire[0].toInt() and 0xFF, wire[1].toInt() and 0xFF,
                wire[2].toInt() and 0xFF, wire.copyOfRange(4, wire.size))
        }
    }
}

File: windows/src/OpenBose.Protocol/BmapPacket.cs
namespace OpenBose.Protocol;

/// <summary>One unframed Bluetooth BMAP packet: block, function, operator, length, payload.</summary>
public sealed class BmapPacket
{
    public const byte GetOperator = 0x01;
    public const byte SetGetOperator = 0x02;
    public const byte StatusOperator = 0x03;
    public const byte ErrorOperator = 0x04;
    public const byte StartOperator = 0x05;

    public byte Block { get; }
    public byte Function { get; }
    public byte Operator { get; }
    private readonly byte[] _payload;
    public ReadOnlyMemory<byte> Payload => _payload;

    public BmapPacket(byte block, byte function, byte op, ReadOnlySpan<byte> payload)
    {
        if (payload.Length > byte.MaxValue) throw new ArgumentOutOfRangeException(nameof(payload));
        Block = block;
        Function = function;
        Operator = op;
        _payload = payload.ToArray();
    }

    public byte[] ToWire()
    {
        byte[] wire = new byte[4 + _payload.Length];
        wire[0] = Block;
        wire[1] = Function;
        wire[2] = Operator;
        wire[3] = (byte)_payload.Length;
        _payload.CopyTo(wire, 4);
        return wire;
    }

    public static BmapPacket ParseExact(ReadOnlySpan<byte> wire)
    {
        if (wire.Length < 4) throw new FormatException("Incomplete BMAP header.");
        int expected = 4 + wire[3];
        if (wire.Length != expected)
            throw new FormatException($"BMAP frame length mismatch: expected {expected}, got {wire.Length}.");
        return new BmapPacket(wire[0], wire[1], wire[2], wire[4..]);
    }
}
